lib/metasploit/framework/compiler/mingw.rb

```diff
@@ -0,0 +1,92 @@
+require 'msf/util/helper'
+require 'open3'
+
+module Metasploit
+  module Framework
+    module Compiler
+      module Mingw
+
+        MINGW_X86 = 'i686-w64-mingw32-gcc'
+        MINGW_X64 = 'x86_64-w64-mingw32-gcc'
+
+        INCLUDE_DIR = File.join(Msf::Config.install_root, 'data', 'headers', 'windows', 'c_payload_util')
+        UTILITY_DIR = File.join(Msf::Config.install_root, 'data', 'utilities', 'encrypted_payload')
+
+        def self.mingw_available?
+          !!(Msf::Util::Helper.which(MINGW_X86) || Msf::Util::Helper.which(MINGW_X64))
+        end
+
+        def self.mingw_x86
+          Msf::Util::Helper.which(MINGW_X86)
+        end
+
+        def self.mingw_x64
+          Msf::Util::Helper.which(MINGW_X64)
+        end
+
+        def self.compile_c(src, opts={})
+          cmd = self.build_cmd(src, opts)
+
+          stdin_err, status = Open3.capture2e(cmd)
+          stdin_err
+        end
+
+        def self.build_cmd(src, opts={})
+          cmd = ''
+          link_options = '-Wl,'
+
+          src_file = opts[:f_name].split('.').first
+          path = File.join(Msf::Config.install_root, "#{src_file}.c")
+          File.write(path, src)
+
+          opt_level = [ 'Os', 'O1', 'O2', 'O3' ].include?(opts[:opt_lvl]) ? "-#{opts[:opt_lvl]} " : "-O2 "
+
+          case opts[:arch]
+          when 'x86'
+            cmd << "#{MINGW_X86} "
+          when 'x64'
+            cmd << "#{MINGW_X64} "
+          else
+            return print_error('Unsupported architecture')
+          end
+
+          cmd << "#{path} -I #{INCLUDE_DIR} "
+          cmd << "-o #{Msf::Config.install_root}/#{opts[:f_name]} "
+
+          # gives each function its own section
+          # allowing them to be reordered
+          cmd << '-ffunction-sections '
+          cmd << '-fno-asynchronous-unwind-tables '
+          cmd << '-nostdlib '
+          cmd << '-fno-ident '
+          cmd << opt_level
+
+          link_options << '--no-seh,'
+          link_options << '-s,' if opts[:strip_symbols]
+          link_options << "-T#{opts[:linker_script]}" if opts[:linker_script]
+
+          cmd << link_options
+
+          cmd
+        end
+
+        def self.cleanup_files(opts={})
+          file_base = opts[:f_name].split('.').first
```

Comment:
Ruby has [`File::Basename`](https://ruby-doc.org/core-2.6.4/File.html#method-c-basename) for this.